docs/running-clusterfuzzlite/gitlab.md

```diff
@@ -0,0 +1,229 @@
+  before_script:
+    # gitlab's container name is not its hostname !
+    - export CFL_CONTAINER_ID=`cut -c9- < /proc/1/cpuset`
+  script:
+    - python3 "/opt/oss-fuzz/infra/cifuzz/cifuzz_combined_entrypoint.py"
+  artifacts:
+    when: always
+    paths:
+      - artifacts/corpus/
+      - artifacts/crashes/
+```
+{% endraw %}
+
+You should then define two [schedules](https://docs.gitlab.com/ee/ci/pipelines/schedules.html)
+In one, you should set the variable `MODE` to `batch` to run the actual batch fuzzing.
+In the other, you should set the variable `MODE` to `prune` for corpus pruning once a day.
+These schedules should target the main/default/`CFL_BRANCH` branch.
+
+### Continuous builds
+
+The continuous build task causes a build to be triggered and uploaded
+whenever a new push is done to main/default branch.
+
+Continuous builds are used when a crash is found during PR fuzzing to determine whether the crash was newly introduced.
+If the crash is not novel, PR fuzzing will not report it.
+This means that there will be fewer unrelated failures when running code change
+fuzzing.
+
+To set up continuous builds, add the following to `.gitlab-ci.yml`:
+
+{% raw %}
+```yaml
+# this name is hardcoded and cannot be changed for gitlab artifacts filestore
+clusterfuzzlite-build:
+  image:
+    name: google/clusterfuzzlite/actions/build_fuzzers@v1
+    entrypoint: [""]
+  stage: fuzz
+  rules:
+    # use $CI_DEFAULT_BRANCH or $CFL_BRANCH
+    - if: $CI_COMMIT_BRANCH == $CFL_BRANCH && $CI_PIPELINE_SOURCE == "push"
+      variables:
+        MODE: "code-change"
+        UPLOAD_BUILD: "true"
+  before_script:
+    # gitlab's container name is not its hostname !
+    - export CFL_CONTAINER_ID=`cut -c9- < /proc/1/cpuset`
+  script:
+    - python3 "/opt/oss-fuzz/infra/cifuzz/cifuzz_combined_entrypoint.py"
+  artifacts:
+    when: always
+    paths:
+      - artifacts/build/
+      - artifacts/crashes/
+```
+{% endraw %}
+
+### Coverage reports
+
+To generate periodic coverage reports, add the following job to
+`.gitlab-ci.yml`:
+
+{% raw %}
+```yaml
+# this name is hardcoded and cannot be changed for gitlab artifacts filestore
+clusterfuzzlite-coverage:
+  image:
+    name: google/clusterfuzzlite/actions/build_fuzzers@v1
+    entrypoint: [""]
+  stage: fuzz
+  variables:
+    SANITIZER: "coverage"
+  rules:
+    - if: $MODE == "coverage"
+  before_script:
+    # gitlab's container name is not its hostname !
+    - export CFL_CONTAINER_ID=`cut -c9- < /proc/1/cpuset`
+  script:
+    - python3 "/opt/oss-fuzz/infra/cifuzz/cifuzz_combined_entrypoint.py"
+  artifacts:
+    when: always
+    paths:
+      - artifacts/coverage/
+      - artifacts/crashes/
+```
+{% endraw %}
+
+You should then define one [schedule](https://docs.gitlab.com/ee/ci/pipelines/schedules.html)
+In it, you should set the variable `MODE` to `coverage`.
+This schedule should target the main/default/`CFL_BRANCH` branch.
+
+
+## Extra configuration
+
+### Gitlab artifacts filestore
```

Comment:
Thanks for going through the trouble of setting this up.
I have some high level comments.
1. This section seems to offer two options: 1. ??? and 2. using  a personal access token.
Can you tell me the tradeoff between the two and we will decide which one to recommend and then we won't even mention the other option.
2. In the github docs we have screenshots for setting up a personal acccess token (https://google.github.io/clusterfuzzlite/running-clusterfuzzlite/github-actions/#storage-repo). If we go that route, can you add screenshots here?
3. Can you add screenshots and explanation for downloading artifacts please?
4. This section makes artifacts seem optional, why should they be optional?
5. The reason we use the git filestore on github in addition to actions is two-fold: 1. it makes browsing coverage reports on the web possible. 2. It makes it possible for two batch fuzzing jobs to save to the corpus at once (side note: @oliverchang: have we ever tested this? Does batch fuzzing force push so that committing works even if its behind HEAD?). Should we do the same on gitlab?

Reply:
re parallel pushing: we auto rebase and retry when push fails to make this work.

Reply:
> Can you tell me the tradeoff between the two and we will decide which one to recommend and then we won't even mention the other option.

The downside of the artifacts is that you need to set up an API token
The downside of the cache is that you should ensure that runners share the cache

I think I gather that you want in the end a `gitlab` filestore that will
- push crashes as jobs artifacts
- push and pull coverage reports in another git repo
- push and pull builds from a cache
- push and pull corpus from a cache

Reply:
Why use cache and artifacts? Why not just one?
Why use git repo for reports? Does gitlab have something like github pages?
Sorry if these are dumb questions/answered elsewhere.